### PR TITLE
Enable Gradient Accumulation fix across all models + trainer fully in forward()

### LINKED ISSUE #34242: Add DDP token averaging for equivalent non-parallel training similar to #34191
### Feature request

Token averaging in gradient accumulation was fixed in #34191 . But token averaging in DDP seems to have the same issue.

---

## Expected behaivor

With all the tokens contributing to loss in each step (in each GPU, gradient accumulation step, and microbatch), the equation becomes:

$$ntokens=\sum\limits_{GPUs} \sum\limits_{gas} \sum\limits_{microb} (label\neq-100)$$

I believe we should average the above tokens at the same time for equivalent non-parallel training.

---

## Current issue

Prior to #34191, the loss/gradients were averaged on $\sum\limits_{GPUs}$, $\sum\limits_{gas}$, and $\sum\limits_{microb}$ separately. And, the introduction of `num_items_in_batch` in #34191 refers to:

$$ntokens=\sum\limits_{gas} \sum\limits_{microb} (label\neq-100)$$

So, the loss/gradients are now averaged on $\sum\limits_{GPUs}$ and $\left(\sum\limits_{gas}\sum\limits_{microb}\right)$ separately. However, this still does not seem equivalent to non-parallel training.

Can we also incorporate $\sum\limits_{GPUs}$ when determining `num_items_in_batch`? Something like `all_reduce(num_items_in_batch)`?

### Motivation

DDP seems not fully equivalent to non-parallel training.

related comments: https://github.com/huggingface/transformers/pull/34191#issuecomment-2421777304

### Your contribution

Found some fairseq implementation of this feature

https://github.com/facebookresearch/fairseq/blob/018621f3cca02ca9de945dc082c3fb1a7f9f2deb/fairseq/trainer.py#L932-L949

### LINKED ISSUE #34263: New GA fix causes training loss multiple times higher across the board (5x to 10x higher)
### System Info

8xH100

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

After updating to the latest master branch of transformer, the training loss is mutiple times higher than before (5x-10x). I tried both SFT and DPO (paired with latest trl master), all having the same problems. 
SFT after GA fix
image

SFT before GA fix
image

### Expected behavior

training loss value should be aligned with old values, or should be expected lower.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Since most users still want OOTB, this trickles the loss kwargs to the rest of the models so that causal loss can be calculated properly 

Fixes # (issue)

Fully fixes https://github.com/huggingface/transformers/issues/34263 / finishes https://github.com/huggingface/transformers/pull/34191 & https://github.com/huggingface/transformers/pull/34198 & fixes https://github.com/huggingface/transformers/issues/34242

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Nice, I don't see Llama being modified, that's probably because it now has `FlashAttentionKwargs` type dict as kwargs. We can create ExtraKwargs, a nested dict with both flash kwargs and loss kwargs and default loss kwargs can be type dict? 
🤗

**Comment by muellerzr** [comment]
@ArthurZucker `LlamaForCausalLM` isn't there bc you did it in your PR 😉

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3610,8 +3612,11 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             labels = inputs.pop("labels")
         else:
             labels = None
-        # if num_items_in_batch is not None:
-        #     inputs["num_items_in_batch"] = num_items_in_batch
+        if self.model_accepts_loss_kwargs:
```
Comment (man-shar) [inline src/transformers/trainer.py:3615]: This if condition doesn't seem to work for `PeftModel` class (it only has `kwargs` not `loss_kwargs` 🫠 )

I tried just changing that condition to `if True` and ran some tests, and the loss calculation worked perfectly for a LORA on a Llama 3 1B.

I'm wondering if there's a safe/non-breaking way to support peft models here as well?

**Comment by man-shar** [comment]
Awesome work by all of you on this. Insane dev speed over the past few days 🙏 🔥

File: src/transformers/trainer.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/trainer.py:3615]: Thanks for the flag, that may be why I couldn't recreate https://github.com/huggingface/transformers/issues/34263#issuecomment-2427711753

Comment (muellerzr) [reply src/transformers/trainer.py:3623]: @man-shar can you try giving peft another go? Should have fixed it

**Review by BenjaminBossan — APPROVED** [review]
I only reviewed the PEFT-related code in `trainer.py` and it LGTM. Thanks Zach.

**Comment by ArthurZucker** [comment]
Rebasing should probably fix quality!

File: src/transformers/trainer.py (hunk 1 above)
Comment (man-shar) [reply src/transformers/trainer.py:3640]: Can confirm `self.model_accepts_loss_kwargs` is the correct value now for a peft model!

But seems like the `compute_loss` function inside trainer.py isn't getting the `num_items_in_batch` argument passed to it from `training_step`.

I notice that argument was removed in 4f3f86d2517e766a5773afb0e58c8bf95df10138 and the commit still says "Experimental". So I assume it will be reverted once you're done experimenting! It should work after that!

Comment (muellerzr) [reply src/transformers/trainer.py:3640]: Yes indeed :)

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2455,7 +2465,7 @@ def _inner_training_loop(
                         self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
                     with self.accelerator.accumulate(model):
-                        tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
+                        tr_loss_step = self.training_step(model, inputs)
```
Comment (zhijian-liu) [inline src/transformers/trainer.py:2468]: Why `num_items_in_batch` is removed here?

File: src/transformers/models/cohere/modeling_cohere.py [hunk 3]
```diff
@@ -1114,6 +1114,7 @@ def forward(
         return_dict: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
         num_logits_to_keep: int = 0,
+        **loss_kwargs,
```
Comment (zhijian-liu) [inline src/transformers/models/cohere/modeling_cohere.py:1117]: I think making it an argument `loss_args` rather than kwargs would be better?

Comment (muellerzr) [reply src/transformers/models/cohere/modeling_cohere.py:1117]: They are explicitly `kwargs`, not positional so kwargs is more accurate

File: src/transformers/trainer.py (hunk 2 above)
Comment (muellerzr) [reply src/transformers/trainer.py]: Leftover artifact while I was experimenting

File: src/transformers/models/cohere/modeling_cohere.py (hunk 3 above)
Comment (zhijian-liu) [reply src/transformers/models/cohere/modeling_cohere.py:1117]: I wasn’t referring to the variable name. My suggestion is to use loss_kwargs instead of **loss_kwargs. The current design prevents forward functions from accepting additional keyword arguments, which could be inconvenient.

Comment (muellerzr) [reply src/transformers/models/cohere/modeling_cohere.py:1117]: We haven't accepted that at all before

Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:1117]: This is prone to evolve either way, with #33932 as an example!

**Comment by whyiug** [comment]
I wonder if the loss function of a multimodal model(e.g. qwen2vl) needs to be fixed.

**Comment by JaheimLee** [comment]
I wonder whether encoder models need to be fixed as well. @ArthurZucker  @muellerzr

**Comment by ArthurZucker** [comment]
Yep! Fill free to open a PR 🤗
